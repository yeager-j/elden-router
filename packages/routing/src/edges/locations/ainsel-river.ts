import { Enemy } from "@workspace/data/enemies";
import { Glitch } from "@workspace/data/glitches";
import { Location } from "@workspace/data/locations";

import { EdgeData } from "#types";

export const ainselRiverEdges: EdgeData[] = [
  // Ainsel Well path
  {
    from: Location.LIURNIA,
    to: Location.AINSEL_RIVER_WELL_DEPTHS,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.AINSEL_RIVER_WELL_DEPTHS,
    to: Location.AINSEL_RIVER_UHL_PALACE_RUINS,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.AINSEL_RIVER_UHL_PALACE_RUINS,
    to: Location.AINSEL_RIVER_DRAGONKIN_ARENA,
    metadata: {
      requirements: {},
    },
  },

  // Ranni's questline path
  {
    from: Location.LIURNIA_THREE_SISTERS,
    to: Location.AINSEL_RIVER_MAIN,
    directed: true,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.LIURNIA_THREE_SISTERS,
    to: Location.AINSEL_RIVER_MAIN,
    directed: true,
    metadata: {
      requirements: {
        requiredGlitch: {
          glitch: Glitch.ZIP,
          description: "",
        },
      },
    },
  },
  {
    from: Location.AINSEL_RIVER_MAIN,
    to: Location.AINSEL_RIVER_NOKSTELLA_ETERNAL_CITY,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.AINSEL_RIVER_NOKSTELLA_ETERNAL_CITY,
    to: Location.AINSEL_RIVER_LAKE_OF_ROT,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.AINSEL_RIVER_LAKE_OF_ROT,
    to: Location.AINSEL_RIVER_GRAND_CLOISTER,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.AINSEL_RIVER_GRAND_CLOISTER,
    to: Location.LIURNIA_MOONLIGHT_ALTAR,
    directed: true,
    metadata: {
      description: "Access via defeating Astel and using the coffin",
      requirements: {
        requiredBosses: [Enemy.ASTEL_NATURALBORN_OF_THE_VOID],
      },
    },
  },
];
