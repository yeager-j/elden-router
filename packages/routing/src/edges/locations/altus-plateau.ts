import {
  Enemy,
  Flag,
  Glitch,
  Location,
  ProgressionItem,
} from "@workspace/data";

import { EdgeData } from "@/types";

export const altusPlateauEdges: EdgeData[] = [
  // Main Altus Plateau access points
  {
    from: Location.LIURNIA,
    to: Location.ALTUS_PLATEAU,
    metadata: {
      description: "Access via Grand Lift of Dectus",
      requirements: [
        {
          type: "item",
          value: ProgressionItem.DECTUS_MEDALLION,
        },
      ],
    },
  },
  {
    from: Location.LIURNIA_RUIN_STREWN_PRECIPICE,
    to: Location.ALTUS_PLATEAU,
    metadata: {
      description: "Alternative path through Ruin-Strewn Precipice",
      requirements: [
        {
          type: "boss",
          value: Enemy.MAGMA_WYRM_MAKAR,
        },
      ],
    },
  },

  // Altus Plateau Dungeons
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_UNSIGHTLY_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_SAINTED_HEROS_GRAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_AURIZA_HEROS_GRAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_AURIZA_SIDE_TOMB,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_WYNDHAM_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },

  // Caves
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_SAGES_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_PERFUMERS_GROTTO,
    metadata: {
      requirements: [],
    },
  },

  // Mines/Tunnels
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_OLD_ALTUS_TUNNEL,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_ALTUS_TUNNEL,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_SEALED_TUNNEL,
    metadata: {
      requirements: [],
    },
  },

  // Other Locations
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_SHADED_CASTLE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS,
    metadata: {
      requirements: [],
    },
  },
  // Mt. Gelmir Connections
  {
    from: Location.LIURNIA_ACADEMY_OF_RAYA_LUCARIA,
    to: Location.MT_GELMIR_VOLCANO_MANOR_INQUISITION_CHAMBER,
    directed: true,
    metadata: {
      description: "Abducted by the Abductor Virgin in Raya Lucaria",
      requirements: [],
    },
  },
  {
    from: Location.MT_GELMIR_VOLCANO_MANOR_INQUISITION_CHAMBER,
    to: Location.MT_GELMIR,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.ABDUCTOR_VIRGINS }],
    },
  },
  {
    from: Location.ALTUS_PLATEAU,
    to: Location.MT_GELMIR,
    metadata: {
      requirements: [],
    },
  },

  // Mt. Gelmir Locations
  {
    from: Location.MT_GELMIR,
    to: Location.MT_GELMIR_VOLCANO_MANOR,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.MT_GELMIR_VOLCANO_MANOR,
    to: Location.MT_GELMIR_VOLCANO_MANOR_POST_EIGLAY,
    metadata: {
      description: "Area after Temple of Eiglay",
      requirements: [{ type: "boss", value: Enemy.GODSKIN_NOBLE }],
    },
  },

  // Mt. Gelmir Dungeons
  {
    from: Location.MT_GELMIR,
    to: Location.MT_GELMIR_HEROS_GRAVE,
    metadata: {
      requirements: [],
    },
  },

  // Mt. Gelmir Caves
  {
    from: Location.MT_GELMIR,
    to: Location.MT_GELMIR_SEETHEWATER_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.MT_GELMIR,
    to: Location.MT_GELMIR_VOLCANO_CAVE,
    metadata: {
      requirements: [],
    },
  },
  // Leyndell Royal Capital Access
  {
    from: Location.ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS,
    to: Location.LEYNDELL_ROYAL_CAPITAL,
    metadata: {
      description: "Enter Leyndell after defeating two Shardbearers",
      requirements: [
        { type: "boss", value: Enemy.DRACONIC_TREE_SENTINEL },
        { type: "boss", value: Enemy.SHARDBEARERS },
        { type: "flag", value: Flag.LEYNDELL_CAPITAL_ASHEN, not: true },
      ],
    },
  },

  // Leyndell Main Progression
  {
    from: Location.LEYNDELL_ROYAL_CAPITAL,
    to: Location.LEYNDELL_ERDTREE_SANCTUARY,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.GODFREY_GOLDEN_SHADE }],
    },
  },
  {
    from: Location.LEYNDELL_ERDTREE_SANCTUARY,
    to: Location.LEYNDELL_ELDEN_THRONE,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.MORGOTT_THE_OMEN_KING }],
    },
  },

  // Leyndell Dungeons and Sub-areas
  {
    from: Location.LEYNDELL_ROYAL_CAPITAL,
    to: Location.LEYNDELL_SUBTERRANEAN_SHUNNING_GROUNDS,
    metadata: {
      requirements: [
        { type: "flag", value: Flag.LEYNDELL_CAPITAL_ASHEN, not: true },
      ],
    },
  },
  {
    from: Location.LEYNDELL_SUBTERRANEAN_SHUNNING_GROUNDS,
    to: Location.LEYNDELL_FRENZIED_FLAME_PROSCRIPTION,
    directed: true,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.MOHG_THE_OMEN }],
    },
  },
  {
    from: Location.LEYNDELL_SUBTERRANEAN_SHUNNING_GROUNDS,
    to: Location.LEYNDELL_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LEYNDELL_ROYAL_CAPITAL,
    to: Location.ALTUS_PLATEAU_FORBIDDEN_LANDS,
    metadata: {
      requirements: [
        { type: "flag", value: Flag.LEYNDELL_CAPITAL_ASHEN, not: true },
      ],
    },
  },
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LEYNDELL_CAPITAL_SUB_AREA,
    directed: true,
    metadata: {
      description: "Use the Teleporter Chest atop the Tower of Return",
      requirements: [],
    },
  },
  {
    from: Location.LEYNDELL_CAPITAL_SUB_AREA,
    to: Location.LEYNDELL_ROYAL_CAPITAL,
    directed: true,
    metadata: {
      requirements: [
        { type: "glitch", value: Glitch.ZIP, description: "" },
        { type: "flag", value: Flag.LEYNDELL_CAPITAL_ASHEN, not: true },
      ],
    },
  },

  // Ashen Capital
  {
    from: Location.ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS,
    to: Location.LEYNDELL_ASHEN_CAPITAL,
    metadata: {
      description: "Capital turns to ash after defeating Maliketh",
      requirements: [{ type: "flag", value: Flag.LEYNDELL_CAPITAL_ASHEN }],
    },
  },
  {
    from: Location.LEYNDELL_ASHEN_CAPITAL,
    to: Location.ALTUS_PLATEAU_FORBIDDEN_LANDS,
    metadata: {
      requirements: [{ type: "flag", value: Flag.LEYNDELL_CAPITAL_ASHEN }],
    },
  },
  {
    from: Location.LEYNDELL_ASHEN_CAPITAL,
    to: Location.LEYNDELL_ASHEN_ELDEN_THRONE,
    metadata: {
      requirements: [
        { type: "boss", value: Enemy.SIR_GIDEON_OFNIR_THE_ALL_KNOWING },
      ],
    },
  },
  {
    from: Location.LEYNDELL_ASHEN_ELDEN_THRONE,
    to: Location.LEYNDELL_FRACTURED_MARIKA,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.GODFREY_HOARAH_LOUX }],
    },
  },
];
