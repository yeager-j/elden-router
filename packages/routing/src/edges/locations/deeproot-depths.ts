import { Enemy, Location } from "@workspace/data";

import { EdgeData } from "@/types";

export const deeprootEdges: EdgeData[] = [
  // Access points to Deeproot Depths
  {
    from: Location.LEYNDELL_FRENZIED_FLAME_PROSCRIPTION,
    to: Location.DEEPROOT_DEPTHS,
    directed: true,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.NOKRON_SIOFRA_AQUEDUCT,
    to: Location.DEEPROOT_DEPTHS,
    directed: true,
    metadata: {
      description: "Access via defeating the Valiant Gargoyles",
      requirements: [{ type: "boss", value: Enemy.VALIANT_GARGOYLES }],
    },
  },

  // Deeproot Depths progression
  {
    from: Location.DEEPROOT_DEPTHS,
    to: Location.DEEPROOT_DEPTHS_NAMELESS_ETERNAL_CITY,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.DEEPROOT_DEPTHS_NAMELESS_ETERNAL_CITY,
    to: Location.DEEPROOT_DEPTHS_FORTISSAX_ARENA,
    metadata: {
      requirements: [],
    },
  },
];
