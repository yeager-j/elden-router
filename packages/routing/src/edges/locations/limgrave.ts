import { Enemy, Location } from "@workspace/data";

import { EdgeData } from "@/types";

export const limgraveEdges: EdgeData[] = [
  {
    from: Location.LIMGRAVE,
    to: Location.ROUNDTABLE_HOLD,
    metadata: {
      description: "Warp to the Roundtable Hold",
      requirements: [],
    },
  },

  // Main Limgrave connections
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_STORMVEIL_CASTLE,
    metadata: {
      description: "Enter Stormveil Castle through the main gate",
      requirements: [{ type: "boss", value: Enemy.MARGIT_THE_FELL_OMEN }],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_WEEPING_PENINSULA,
    metadata: {
      requirements: [],
    },
  },

  // Stormveil Castle sub-locations
  {
    from: Location.LIMGRAVE_STORMVEIL_CASTLE,
    to: Location.LIMGRAVE_STORMVEIL_CASTLE_THRONE_ROOM,
    metadata: {
      description: "Access throne room after navigating castle",
      requirements: [{ type: "boss", value: Enemy.GODRICK_THE_GRAFTED }],
    },
  },
  {
    from: Location.LIMGRAVE_STORMVEIL_CASTLE,
    to: Location.LIMGRAVE_DIVINE_TOWER,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIMGRAVE_STORMVEIL_CASTLE_THRONE_ROOM,
    metadata: {
      description: "Backtrack from Liurnia",
      requirements: [],
    },
  },

  // Limgrave Dungeons
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_STORMFOOT_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_MURKWATER_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_DEATHTOUCHED_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },

  // Limgrave Caves
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_GROVESIDE_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_COASTAL_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_MURKWATER_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_HIGHROAD_CAVE,
    metadata: {
      requirements: [],
    },
  },

  // Limgrave Mines
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_LIMGRAVE_TUNNELS,
    metadata: {
      requirements: [],
    },
  },

  // Special Locations
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_WAYPOINT_RUINS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_WAYPOINT_RUINS,
    to: Location.LIMGRAVE_WAYPOINT_RUINS_SELLEN_ROOM,
    metadata: {
      description: "Access Sellen's underground study",
      requirements: [{ type: "boss", value: Enemy.MAD_PUMPKIN_HEAD }],
    },
  },
  {
    from: Location.LIMGRAVE,
    to: Location.LIMGRAVE_FRINGEFOLK_HEROS_GRAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_COASTAL_CAVE,
    to: Location.LIMGRAVE_CHURCH_OF_DRAGON_COMMUNION,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.DEMI_HUMAN_CHIEF }],
    },
  },
  {
    from: Location.LIURNIA_FOUR_BELFRIES,
    to: Location.CHAPEL_OF_ANTICIPATION,
    metadata: {
      requirements: [],
    },
  },

  // Weeping Peninsula sub-locations
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LIMGRAVE_CASTLE_MORNE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LIMGRAVE_MORNE_TUNNEL,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LIMGRAVE_EARTHBORE_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LIMGRAVE_IMPALERS_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LIMGRAVE_TOMBSWARD_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIMGRAVE_WEEPING_PENINSULA,
    to: Location.LIMGRAVE_TOMBSWARD_CAVE,
    metadata: {
      requirements: [],
    },
  },
];
