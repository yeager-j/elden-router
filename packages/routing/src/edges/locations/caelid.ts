import { Enemy } from "@workspace/data/enemies";
import { Flag } from "@workspace/data/flags";
import { Location } from "@workspace/data/locations";

import { EdgeData } from "#types";

export const caelidEdges: EdgeData[] = [
  // Main access to Caelid
  {
    from: Location.LIMGRAVE,
    to: Location.CAELID,
    metadata: {
      requirements: {},
    },
  },

  // Main Caelid connections
  {
    from: Location.CAELID,
    to: Location.CAELID_REDMANE_CASTLE,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.CAELID,
    to: Location.CAELID_REDMANE_CASTLE_FESTIVAL,
    metadata: {
      description: "Castle transforms into festival grounds",
      requirements: {
        requiredEnabledFlags: [Flag.RADAHN_FESTIVAL_ENABLED],
      },
    },
  },
  {
    from: Location.CAELID,
    to: Location.DRAGONBARROW,
    metadata: {
      requirements: {},
    },
  },

  // Festival and Radahn's arena
  {
    from: Location.CAELID_REDMANE_CASTLE_FESTIVAL,
    to: Location.CAELID_WAILING_DUNES,
    metadata: {
      description: "Arena where you fight Radahn",
      requirements: {},
    },
  },
  {
    from: Location.CAELID_WAILING_DUNES,
    to: Location.CAELID_WAR_DEAD_CATACOMBS,
    metadata: {
      description: "Only accessible after defeating Radahn",
      requirements: {
        requiredBosses: [Enemy.STARSCOURGE_RADAHN],
      },
    },
  },

  // Dungeons
  {
    from: Location.CAELID,
    to: Location.CAELID_MINOR_ERDTREE_CATACOMBS,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.CAELID,
    to: Location.CAELID_CAELID_CATACOMBS,
    metadata: {
      requirements: {},
    },
  },

  // Caves
  {
    from: Location.CAELID,
    to: Location.CAELID_ABANDONED_CAVE,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.CAELID,
    to: Location.CAELID_GAOL_CAVE,
    metadata: {
      requirements: {},
    },
  },

  // Tunnels
  {
    from: Location.CAELID,
    to: Location.CAELID_GAEL_TUNNEL,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.CAELID,
    to: Location.CAELID_SELLIA_CRYSTAL_TUNNEL,
    metadata: {
      requirements: {},
    },
  },

  // Special Areas
  {
    from: Location.DRAGONBARROW,
    to: Location.CAELID_SELLIA_HIDEAWAY,
    metadata: {
      description: "Access via a Spiritspring in Dragonbarrow",
      requirements: {},
    },
  },

  // Dragonbarrow sub-locations
  {
    from: Location.DRAGONBARROW,
    to: Location.DRAGONBARROW_CAVE,
    metadata: {
      requirements: {},
    },
  },
  {
    from: Location.DRAGONBARROW,
    to: Location.DRAGONBARROW_DIVINE_TOWER_OF_CAELID,
    metadata: {
      requirements: {},
    },
  },
];
