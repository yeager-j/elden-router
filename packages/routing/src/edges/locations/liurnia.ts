import { Enemy } from "@workspace/data/enemies";
import { Glitch } from "@workspace/data/glitches";
import { ProgressionItem } from "@workspace/data/items";
import { Location } from "@workspace/data/locations";

import { EdgeData } from "#types";

export const liurniaEdges: EdgeData[] = [
  // Main access to Liurnia
  {
    from: Location.LIMGRAVE,
    to: Location.LIURNIA,
    metadata: {
      description: "Take the shortcut near Stormveil Castle",
      requirements: [],
    },
  },

  // Main Liurnia connections
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_ACADEMY_OF_RAYA_LUCARIA,
    metadata: {
      description:
        "Enter through the Academy Main Gate using the Glintstone Key",
      requirements: [
        {
          type: "item",
          value: ProgressionItem.ACADEMY_GLINTSTONE_KEY,
        },
      ],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_CARIA_MANOR,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_FOUR_BELFRIES,
    metadata: {
      requirements: [],
    },
  },

  // Raya Lucaria Academy sub-locations
  {
    from: Location.LIURNIA_ACADEMY_OF_RAYA_LUCARIA,
    to: Location.LIURNIA_RAYA_LUCARIA_ROOFTOPS,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.RED_WOLF_OF_RADAGON }],
    },
  },
  {
    from: Location.LIURNIA_RAYA_LUCARIA_ROOFTOPS,
    to: Location.LIURNIA_RAYA_LUCARIA_GRAND_LIBRARY,
    metadata: {
      requirements: [],
    },
  },

  // Three Sisters area (accessed through Caria Manor)
  {
    from: Location.LIURNIA_CARIA_MANOR,
    to: Location.LIURNIA_THREE_SISTERS,
    metadata: {
      description: "Access through Caria Manor's upper level",
      requirements: [{ type: "boss", value: Enemy.ROYAL_KNIGHT_LORETTA }],
    },
  },
  {
    from: Location.LIURNIA_CARIA_MANOR,
    to: Location.LIURNIA_THREE_SISTERS,
    metadata: {
      requirements: [{ type: "glitch", value: Glitch.ZIP, description: "" }],
    },
  },

  // Carian Study Hall and Divine Tower
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_CARIAN_STUDY_HALL,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_CARIAN_STUDY_HALL_INVERTED,
    metadata: {
      description: "Use the Carian Inverted Statue to flip the study hall",
      requirements: [
        {
          type: "item",
          value: ProgressionItem.CARIAN_INVERTED_STATUE,
        },
      ],
    },
  },
  {
    from: Location.LIURNIA_CARIAN_STUDY_HALL_INVERTED,
    to: Location.LIURNIA_DIVINE_TOWER,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.GODSKIN_NOBLE }],
    },
  },

  // Dungeons
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_CLIFFBOTTOM_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_ROADS_END_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_BLACK_KNIFE_CATACOMBS,
    metadata: {
      requirements: [],
    },
  },

  // Caves
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_STILLWATER_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_LAKESIDE_CRYSTAL_CAVE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA_LAKESIDE_CRYSTAL_CAVE,
    to: Location.LIURNIA_SLUMBERING_WOLFS_SHACK,
    metadata: {
      requirements: [{ type: "boss", value: Enemy.BLOODHOUND_KNIGHT }],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_ACADEMY_CRYSTAL_CAVE,
    metadata: {
      requirements: [],
    },
  },

  // Other locations
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_RAYA_LUCARIA_CRYSTAL_TUNNEL,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_RUIN_STREWN_PRECIPICE,
    metadata: {
      requirements: [],
    },
  },
  {
    from: Location.LIURNIA,
    to: Location.LIURNIA_MOONLIGHT_ALTAR,
    metadata: {
      requirements: [{ type: "glitch", value: Glitch.ZIP, description: "" }],
    },
  },
];
