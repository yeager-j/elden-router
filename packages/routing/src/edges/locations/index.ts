import { ainselRiverEdges } from "#edges/locations/ainsel-river";
import { altusPlateauEdges } from "#edges/locations/altus-plateau";
import { caelidEdges } from "#edges/locations/caelid";
import { deeprootEdges } from "#edges/locations/deeproot-depths";
import { limgraveEdges } from "#edges/locations/limgrave";
import { liurniaEdges } from "#edges/locations/liurnia";
import {
  farumAzulaEdges,
  mountaintopsEdges,
  snowfieldHaligtreeEdges,
} from "#edges/locations/mountaintops";
import { siofraRiverEdges } from "#edges/locations/siofra-river";

export const allLocations = [
  ...limgraveEdges,
  ...liurniaEdges,
  ...caelidEdges,
  ...altusPlateauEdges,
  ...mountaintopsEdges,
  ...farumAzulaEdges,
  ...snowfieldHaligtreeEdges,
  ...siofraRiverEdges,
  ...ainselRiverEdges,
  ...deeprootEdges,
];
