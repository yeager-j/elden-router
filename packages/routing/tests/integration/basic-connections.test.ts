import { describe, expect, test } from "vitest";

import { Location } from "@workspace/data/locations";

import { getPathToDestination } from "#graph";
import { PathSettings } from "#types";

describe("basic connections", () => {
  const config: PathSettings = {
    allowBosses: false,
    allowGlitches: false,
    acquiredItems: new Set(),
    completedQuestlineStages: new Set(),
    flagsEnabled: new Set(),
  };

  test("Limgrave -> Stormveil Castle", () => {
    const route = getPathToDestination(
      Location.LIMGRAVE_STORMVEIL_CASTLE,
      config,
    );

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Waypoint Ruins", () => {
    const route = getPathToDestination(
      Location.LIMGRAVE_WAYPOINT_RUINS,
      config,
    );

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Liurnia", () => {
    const route = getPathToDestination(Location.LIURNIA, config);

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Caelid", () => {
    const route = getPathToDestination(Location.CAELID, config);

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Divine Tower of Caelid", () => {
    const route = getPathToDestination(
      Location.DRAGONBARROW_DIVINE_TOWER_OF_CAELID,
      config,
    );

    expect(route).toMatchSnapshot();
  });
});
