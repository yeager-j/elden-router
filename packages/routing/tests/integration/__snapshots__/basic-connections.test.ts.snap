// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic connections > Limgrave -> Caelid 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "requirements": {},
      },
      "to": "CAELID",
    },
  ],
  "reasons": [],
}
`;

exports[`basic connections > Limgrave -> Divine Tower of Caelid 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "requirements": {},
      },
      "to": "CAELID",
    },
    {
      "from": "CAELID",
      "metadata": {
        "requirements": {},
      },
      "to": "DRAGONBARROW",
    },
    {
      "from": "DRAGONBARROW",
      "metadata": {
        "requirements": {},
      },
      "to": "DRAGONBARROW_DIVINE_TOWER_OF_CAELID",
    },
  ],
  "reasons": [],
}
`;

exports[`basic connections > Limgrave -> Liurnia 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "description": "Take the shortcut near Stormveil Castle",
        "requirements": {},
      },
      "to": "LIURNIA",
    },
  ],
  "reasons": [],
}
`;

exports[`basic connections > Limgrave -> Stormveil Castle 1`] = `
{
  "isError": true,
  "pathSteps": null,
  "reasons": [
    "Killing Margit, the Fell Omen is required, but you disallowed fighting bosses.",
  ],
}
`;

exports[`basic connections > Limgrave -> Waypoint Ruins 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "requirements": {},
      },
      "to": "LIMGRAVE_WAYPOINT_RUINS",
    },
  ],
  "reasons": [],
}
`;
